packages/flutter/lib/src/widgets/binding.dart

```diff
@@ -158,9 +158,9 @@ abstract class WidgetsBinding extends BindingBase implements GestureBinding, Ren
   /// [WidgetsApp] uses this in conjunction with a [Navigator] to
   /// cause the back button to close dialog boxes, return from modal
   /// pages, and so forth.
-  void handlePopRoute() {
+  Future<Null> handlePopRoute() async {
     for (WidgetsBindingObserver observer in _observers) {
-      if (observer.didPopRoute())
+      if (await observer.didPopRoute())
```

Comment:
we probably need to take a local copy of _observers here or otherwise handle what happens while this now long-running API gets fiddled with while in flight. We used to have the rule that didPopRoute couldn't change the observers but that no longer works since it's entirely async and minutes could pass.